### PR TITLE
Unify  <type>_up_ref methods returned type

### PR DESCRIPTION
**Pull request by FdaSilvaYY**
Related to a discussion in #818 .
Add  status return value for some of them without it. 
Update the docs.
Are this PR compatible with API freeze cut-off ?

### CODE REVIEW

File: doc/crypto/EVP_PKEY_new.pod [hunk 1]
```diff
@@ -37,7 +37,7 @@ used.
 EVP_PKEY_new() returns either the newly allocated B<EVP_PKEY> structure or
 B<NULL> if an error occurred.
 
-EVP_PKEY_up_ref() and EVP_PKEY_free() do not return a value.
+EVP_PKEY_free() do not return a value.
```
Comment (richsalz) [inline doc/crypto/EVP_PKEY_new.pod]: "does not return"

File: doc/crypto/X509_new.pod [hunk 2]
```diff
@@ -46,7 +46,7 @@ If the allocation fails, X509_new() returns B<NULL> and sets an error
 code that can be obtained by L<ERR_get_error(3)>.
 Otherwise it returns a pointer to the newly allocated structure.
 
-X509_free() and X509_up_ref() do not return a value.
+X509_free() do not return a value.
```
Comment (richsalz) [inline doc/crypto/X509_new.pod]: "does not return"  but is that even needed for void functions?  Let's just remove those lines.

**Comment by richsalz** [comment]
thanks! +1 and will merge after another review.

File: crypto/evp/p_lib.c [hunk 3]
```diff
@@ -205,10 +205,14 @@ EVP_PKEY *EVP_PKEY_new(void)
     return ret;
 }
 
-void EVP_PKEY_up_ref(EVP_PKEY *pkey)
+int EVP_PKEY_up_ref(EVP_PKEY *pkey)
 {
     int i;
     CRYPTO_atomic_add(&pkey->references, 1, &i, pkey->lock);
```
Comment (richsalz) [inline crypto/evp/p_lib.c]: those can be bugfixes after the API freeze.  but yes, shoudl happen.

Comment (FdaSilvaYY) [inline crypto/evp/p_lib.c]: Done only on already modified code.
It's deserves a larger and separated review/commit to fix them all.

Comment (ghedo) [inline crypto/evp/p_lib.c]: Are there currently any _up_ref() with the int return that don't check for errors? I think I already converted them all during the threading conversion (though I guess I may have missed some).

**Comment by FdaSilvaYY** [comment]
Rebased on today HEAD, it builds OK.
Just need one more review.

File: doc/crypto/EVP_PKEY_new.pod [hunk 4]
```diff
@@ -37,8 +37,6 @@ used.
 EVP_PKEY_new() returns either the newly allocated B<EVP_PKEY> structure or
 B<NULL> if an error occurred.
 
-EVP_PKEY_up_ref() and EVP_PKEY_free() do not return a value.
```
Comment (mattcaswell) [inline doc/crypto/EVP_PKEY_new.pod]: A description of what the return value from EVP_PKEY_up_ref() means is required here.

File: doc/crypto/X509_new.pod [hunk 5]
```diff
@@ -46,8 +46,6 @@ If the allocation fails, X509_new() returns B<NULL> and sets an error
 code that can be obtained by L<ERR_get_error(3)>.
 Otherwise it returns a pointer to the newly allocated structure.
 
-X509_free() and X509_up_ref() do not return a value.
```
Comment (mattcaswell) [inline doc/crypto/X509_new.pod]: Describe X509_up_ref() return value meaning

File: doc/ssl/SSL_CTX_new.pod [hunk 6]
```diff
@@ -17,7 +17,7 @@ functions
  #include <openssl/ssl.h>
 
  SSL_CTX *SSL_CTX_new(const SSL_METHOD *method);
- void SSL_CTX_up_ref(SSL_CTX *ctx);
+ int SSL_CTX_up_ref(SSL_CTX *ctx);
```
Comment (mattcaswell) [inline doc/ssl/SSL_CTX_new.pod]: Please add to return values section

**Comment by mattcaswell** [comment]
Probably needs a CHANGES entry. As this is only return value changes I *think* this is ok to go in??

**Comment by FdaSilvaYY** [comment]
Docs updated.
This change don't value one more CHANGES entry, IMHO.
But if you insist...

**Comment by mattcaswell** [comment]
Updates look fine, but I do think a CHANGES entry is warranted. The reason is that it is a change to an existing API which is likely to go unnoticed if we don't draw attention to it.

**Comment by levitte** [comment]
Looks fine to me.  @mattcaswell?

File: CHANGES [hunk 7]
```diff
@@ -4,6 +4,13 @@
 
  Changes between 1.0.2g and 1.1.0  [xx XXX xxxx]
 
+  *) Unify TYPE_up_ref(obj) methods signature.
+     SSL_CTX_up_ref(), SSL_up_ref(), X509_up_ref(), EVP_PKEY_up_ref(), 
+     X509_CRL_up_ref(), X509_OBJECT_up_ref_count() methods are now returning an
+     int (instead of void) like all others TYPE_up_ref() methods.
+     So now these methods also check the return value of CRYPTO_atomic_add(), 
+     and the validity of object reference counter.
```
Comment (mattcaswell) [inline CHANGES]: Nit: Normally you would add your name in square brackets underneath the CHANGES entry (although I note the one below this doesn't have that either....but all the others do!).

**Comment by mattcaswell** [comment]
Subject to the nit being fixed I can +1 this. @richsalz there's been a few changes since you added the "reviewed" label. Can you reconfirm this?

**Comment by richsalz** [comment]
Wholeheartedly yes, +1 remains.

**Comment by richsalz** [comment]
+1

**Comment by FdaSilvaYY** [comment]
Ping @mattcaswell : is it ok for you ?

**Comment by mattcaswell** [comment]
Merged.

**Pull request closed without merging**